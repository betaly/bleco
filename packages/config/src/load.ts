import {Configuration} from '@boost/config';
import popu from 'popu';
import {Config} from './config';
import {Env} from './env';
import {toArray} from 'tily/array/toArray';
import {AnyObj} from 'tily/typings/types';

/**
 * Load config from dirs in overwriting order.
 *
 * @param name
 * @param fromDirs
 */

export async function load<T extends object>(name: string | Configuration<T>, fromDirs?: string | string[]) {
  const c = name instanceof Configuration ? name : new Config<T>(name);
  const dirs = toArray(fromDirs ?? process.cwd());
  const {env} = Env.load(dirs, {mergeToProcessEnv: false});
  const config = await loadConfigs(c, dirs);
  return popu(config, env);
}

async function loadConfigs<T extends object>(c: Configuration<T>, fromDirs: string[]) {
  const answer: AnyObj = {};
  for (const dir of fromDirs) {
    const {config} = await c.loadConfigFromRoot(dir);
    Object.assign(answer, config);
  }
  return answer;
}
