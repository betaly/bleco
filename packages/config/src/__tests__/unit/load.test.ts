import {load} from '../../load';
import {fixturePath} from '../support';

describe('load config', function () {
  const oldEnv = process.env;

  beforeEach(() => {
    jest.resetModules(); // Most important - it clears the cache
    process.env = {...oldEnv}; // Make a copy
  });

  afterAll(() => {
    process.env = oldEnv; // Restore old environment
  });

  it('should load from current working directory', async () => {
    const cwd = process.cwd();
    process.chdir(fixturePath('config-basic'));
    const config = await load('demo');
    expect(config).toMatchSnapshot();
    process.chdir(cwd);
  });

  it('should load from specific directory', async () => {
    const config = await load('demo', fixturePath('config-basic'));
    expect(config).toMatchSnapshot();
  });

  it('should return empty config loading from directory without config file and environments', async () => {
    const config = await load('demo', '.');
    expect(config).toEqual({});
  });

  it('should not merge env variable to process.env with mergeToProcessEnv disabled', async () => {
    const config = await load('demo', fixturePath('config-basic'));
    expect(process.env.FOO_BAR_A).toBeUndefined();
  });

  it('should merge env variable to process.env with mergeToProcessEnv enabled', async () => {
    const config = await load('demo', fixturePath('config-basic'), {
      mergeToProcessEnv: true,
    });
    expect(process.env.FOO_BAR_A).toEqual('foo_bar_3');
  });

  it('should load with defaults config', async () => {
    const config = await load('demo', fixturePath('config-basic'), {
      defaults: {
        foo: 'bar',
        // The value ends with '*' will be treated as original value
        password: '123456*',
      },
    });
    expect(config).toMatchSnapshot();
  });

  it('should remove null and blank values default', async () => {
    const config = await load('demo', fixturePath('config-basic'), {
      defaults: {
        blank: '',
        null: null,
        awesome: 'yes',
      },
    });
    expect(config).toMatchSnapshot();
  });

  it('should load from multiple directories with overriding orders', async () => {
    const config = await load('demo', [fixturePath('config-default'), fixturePath('config-override')]);
    expect(config).toMatchSnapshot();
  });

  it('should throw error loading from none existing directory', async () => {
    await expect(load('demo', '/none-exists')).rejects.toThrow(/no such file or directory/);
  });
});
