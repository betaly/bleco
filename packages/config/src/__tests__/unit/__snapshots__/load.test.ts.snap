// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load config should load from current working directory 1`] = `
{
  "deep": {
    "object": {
      "property": "value",
    },
  },
  "foo": {
    "bar": {
      "a": "foo_bar_3",
      "b": "foo_bar_2",
      "c": "foo_bar_4",
      "d": "\${FOO_BAR_D}",
    },
  },
  "name": "hello",
}
`;

exports[`load config should load from multiple directories with overriding orders 1`] = `
{
  "foo": "bar",
  "var1": "C",
}
`;

exports[`load config should load from specific directory 1`] = `
{
  "deep": {
    "object": {
      "property": "value",
    },
  },
  "foo": {
    "bar": {
      "a": "foo_bar_3",
      "b": "foo_bar_2",
      "c": "foo_bar_4",
      "d": "\${FOO_BAR_D}",
    },
  },
  "name": "hello",
}
`;
